README.md
React-hint
==========

[![npm package][npm-badge]][npm] [![npm package][npm-downloads]][npm]

**React-hint** is a small tooltip component for [React](https://github.com/facebook/react) which is developed with simplicity and performance in mind. It also plays nicely with [Preact](https://github.com/developit/preact) and [Inferno](https://github.com/trueadm/inferno).

![react-hint tooltip](https://raw.githubusercontent.com/slmgc/react-hint/master/demo/react-hint.gif)
![custom tooltip](https://raw.githubusercontent.com/slmgc/react-hint/master/demo/custom-tooltip.png)

How to install
--------------
```
npm i -S react-hint
```

How to import
-------------

```jsx
// React
import React from 'react'
import ReactHintFactory from 'react-hint'
const ReactHint = ReactHintFactory(React)

// Preact
import {h, Component, createRef} from 'preact'
import ReactHintFactory from 'react-hint'
const ReactHint = ReactHintFactory({createElement: h, Component, createRef})

// Inferno
import Inferno from 'inferno-compat'
import ReactHintFactory from 'react-hint'
const ReactHint = ReactHintFactory(Inferno)

// UMD
const ReactHint = window.ReactHintFactory(window.React)
```

You don't need to include ReactHint in every component which uses tooltips, just include it once in the topmost container component.

Use
---

ReactHint is (in 99% of cases) a singleton-component which is used to render tooltips which appear on multiple elements :
 
```jsx
<ReactHint autoPosition events />
<button data-rh="tooltip 1">Hover me 1 !</button>
<button data-rh="tooltip 2">Hover me 2 !</button>
<button data-rh="tooltip 3">Hover me 3 !</button>
<button data-rh="tooltip 4">Hover me 4 !</button>
```

The text content which appears inside the tooltip is set by `data-rh` attribute.

Tooltip will appear on hover on every element with `data-rh` attribute.

*Note : tooltip can also be toggled by calling `toggleHint()` on the ref of a the component:*
 
```jsx
<ReactHint autoPosition events ref={(ref) => { this.tooltip = ref; }} />
<div data-rh="tooltip">Element with tooltip</div>
<button onClick={() => { this.tooltip.toggleHint(); }}>Click me to toggle React Hint on element !</button>
```

### custom content

In case you need to define custom content (HTML), you must use the `onRenderContent` prop of ReactHint:

```jsx
<ReactHint
	autoPosition
	events
	onRenderContent={(target) => (<div><p>`tooltip ${target.number}`</p></div>)}
/>
<button data-rh data-number="1">Hover me 1 !</button>
<button data-rh data-number="2">Hover me 2 !</button>
<button data-rh data-number="3">Hover me 3 !</button>
<button data-rh data-number="4">Hover me 4 !</button>
```

Use `data-abcdef` attribute on component which uses tooltip to pass data which can be accessed via `target.abcdef` in `onRenderContent()`.

*ReactHint is not your regular wrapping tooltip component library, e.g. this is **incorrect**:*

```jsx
<ReactHint>
	<div>Content of the tooltip</div>
</ReactHint>
```

### multiple instances

In case you need to define multiple instances of `ReactHint` (ex to show tooltips with different content layout), you can customize the attribute name per instance.
ReactHint also supports custom tooltip content with attached event handlers by overriding the content renderer and returning a react node.

```jsx

// default tooltip
<ReactHint autoPosition events />

// custom tooltip 1
<ReactHint
	persist
	attribute="data-custom-1"
	events={{click: true}}
	onRenderContent={(target) => (<div><p>`tooltip ${target.number}`</p></div>)}
/>

// custom tooltip 2
<ReactHint
	persist
	attribute="data-custom-2"
	events={{click: true}}
	onRenderContent={(target) => (<h1>`tooltip ${target.title}`</h1>)}
/>

<button data-rh="default tooltip 1">Hover me 1 to show default tooltip !</button>
<button data-rh="default tooltip 2">Hover me 2 to show default tooltip !</button>

<button data-custom-1	data-custom-1-number="123">Hover me to show custom tooltip 1 !</button>
<button data-custom-1	data-custom-1-number="456">Hover me to show custom tooltip 1 !</button>

<button data-custom-2	data-custom-2-title="Hello">Hover me to show custom tooltip 2 !</button>
<button data-custom-2	data-custom-2-title="World">Hover me to show custom tooltip 2 !</button>
```

*Note : when using custom attribute name, data should be passed prefixed with attribute name as shown above.*

Options
-------

| ReactHint Property | Type                                                        | Default Value | Description
| :----------------- | :---------------------------------------------------------- | :------------ | :----------
| attribute          | String                                                      | "data-rh"     | Allows setting a custom tooltip attribute instead of the default one.
| autoPosition       | Boolean                                                     | false         | Autopositions tooltips based on closeness to window borders.
| className          | String                                                      | "react-hint"  | You can override the tooltip style by passing the `className` property.
| delay              | Number or {show: Number, hide: Number}                      | 0             | The default delay (in milliseconds) before showing/hiding the tooltip.
| events             | Boolean or {click: Boolean, focus: Boolean, hover: Boolean} | false         | Enables/disables all events or a subset of events.
| onRenderContent    | Function                                                    |               | Allows rendering of custom HTML content (with attached event handlers). Pass a function which returns a react node.
| persist            | Boolean                                                     | false         | Hide the tooltip only on outside click, hover, etc.
| position           | "top", "left", "right", "bottom"                            | "top"         | Allows setting the default tooltip placement.
| ref                | Function                                                    |               | You can pass a function which will get a reference to the tooltip instance.

| DOM Element Attribute | Type                             | Default Value | Description
| :-------------------- | :------------------------------- | :------------ | :----------
| data-rh               | String                           |               | Sets the tooltip's text content (if `onRenderContent` is not used to set custom HTML content).
| data-rh-at            | "top", "left", "right", "bottom" | "top"         | Allows overriding the default tooltip placement.

Full Example
------------

```jsx
import React from 'react'
import {render} from 'react-dom'
import ReactHintFactory from 'react-hint'
import 'react-hint/css/index.css'

const ReactHint = ReactHintFactory(React)
class App extends React.Component {
	onRenderContent = (target, content) => {
		const {catId} = target.dataset
		const width = 240
		const url = `https://images.pexels.com/photos/${catId}/pexels-photo-${catId}.jpeg?w=${width}`

		return <div className="custom-hint__content">
			<img src={url} width={width} />
			<button ref={(ref) => ref && ref.focus()}
				onClick={() => this.instance.toggleHint()}>Ok</button>
		</div>
	}

	render() {
		return <div>
			<ReactHint autoPosition events delay={{show: 100, hide: 1000}} />
			<ReactHint persist
				attribute="data-custom"
				className="custom-hint"
				events={{click: true}}
				onRenderContent={this.onRenderContent}
				ref={(ref) => this.instance = ref} />

			<button data-rh="Default">Default</button>
			<button data-rh="Top" data-rh-at="top">Top</button>
			<button data-rh="Right" data-rh-at="right">Right</button>
			<button data-rh="Bottom" data-rh-at="bottom">Bottom</button>
			<button data-rh="Left" data-rh-at="left">Left</button>

			<button data-custom
				data-custom-at="bottom"
				data-cat-id="10913">Click Me</button>

			<button data-custom
				data-custom-at="bottom"
				data-cat-id="416088">Click Me</button>
		</div>
	}
}

render(<App />, demo)
```

License
-------
MIT

[npm-badge]: https://img.shields.io/npm/v/react-hint.png
[npm-downloads]: https://img.shields.io/npm/dm/react-hint.svg
[npm]: https://www.npmjs.org/package/react-hint


CHANGELOG.md
## 3.2.1 (April 2, 2021)
* #49: Fix ininite re-renders during tooltip positioning

## 3.2.0 (December 5, 2018)
* #29: Allow setting different show & hide delay

## 3.1.1 (August 2, 2018)
* #28: Make react-hint ARIA-compliant

## 3.1.0 (May 12, 2018)
* #25: Add support for tooltip autopositioning via `autoPosition` property

## 3.0.1 (Mar 5, 2018)
* #19: Fix infinite setState loop in IE11

## 3.0.0 (Nov 4, 2017)
* Add support for custom content rendering via `onRenderContent` property
* Add support for `click`, `focus` and `hover` in `events` property
* Export `ReactHintFactory` function via the default export
* Depreacte templates via `#element-id` in `data-rh` in favor of custom content rendering
* Rename `hover` property to `persist`

## 2.0.3 (Sep 30, 2017)
* #12: Fix iOS mouseover

## 2.0.2 (Sep 6, 2017)
* #9: Fix production build in create-react-app

## 2.0.1 (Aug 21, 2017)
* Add missing build artefacts

## 2.0.0 (Aug 21, 2017)
* Add support for multiple `<ReactHint />` instances
* Add support for `attribute` property instead of the default `data-rh`
* Add support for `delay` property instead of the default `100ms` delay
* Add support for `events` property to enable/disable mouse hover events
* Add support for `hover` property to enable/disable the tooltip hovering
* Export `ReactHintFactory` function instead of the default export
* Replace `ReactHint` with `ReactHintFactory` for UMD builds
* Deprecate `ReactHint.instance` property
* Deprecate `data-rh-cls` attribute


umd/react-hint.js
/*!
 * react-hint v3.2.1 - https://react-hint.js.org
 * MIT Licensed
 */
(function webpackUniversalModuleDefinition(root, factory) {
	if(typeof exports === 'object' && typeof module === 'object')
		module.exports = factory();
	else if(typeof define === 'function' && define.amd)
		define([], factory);
	else if(typeof exports === 'object')
		exports["ReactHintFactory"] = factory();
	else
		root["ReactHintFactory"] = factory();
})(window, function() {
return /******/ (function(modules) { // webpackBootstrap
/******/ 	// The module cache
/******/ 	var installedModules = {};
/******/
/******/ 	// The require function
/******/ 	function __webpack_require__(moduleId) {
/******/
/******/ 		// Check if module is in cache
/******/ 		if(installedModules[moduleId]) {
/******/ 			return installedModules[moduleId].exports;
/******/ 		}
/******/ 		// Create a new module (and put it into the cache)
/******/ 		var module = installedModules[moduleId] = {
/******/ 			i: moduleId,
/******/ 			l: false,
/******/ 			exports: {}
/******/ 		};
/******/
/******/ 		// Execute the module function
/******/ 		modules[moduleId].call(module.exports, module, module.exports, __webpack_require__);
/******/
/******/ 		// Flag the module as loaded
/******/ 		module.l = true;
/******/
/******/ 		// Return the exports of the module
/******/ 		return module.exports;
/******/ 	}
/******/
/******/
/******/ 	// expose the modules object (__webpack_modules__)
/******/ 	__webpack_require__.m = modules;
/******/
/******/ 	// expose the module cache
/******/ 	__webpack_require__.c = installedModules;
/******/
/******/ 	// define getter function for harmony exports
/******/ 	__webpack_require__.d = function(exports, name, getter) {
/******/ 		if(!__webpack_require__.o(exports, name)) {
/******/ 			Object.defineProperty(exports, name, { enumerable: true, get: getter });
/******/ 		}
/******/ 	};
/******/
/******/ 	// define __esModule on exports
/******/ 	__webpack_require__.r = function(exports) {
/******/ 		if(typeof Symbol !== 'undefined' && Symbol.toStringTag) {
/******/ 			Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });
/******/ 		}
/******/ 		Object.defineProperty(exports, '__esModule', { value: true });
/******/ 	};
/******/
/******/ 	// create a fake namespace object
/******/ 	// mode & 1: value is a module id, require it
/******/ 	// mode & 2: merge all properties of value into the ns
/******/ 	// mode & 4: return value when already ns object
/******/ 	// mode & 8|1: behave like require
/******/ 	__webpack_require__.t = function(value, mode) {
/******/ 		if(mode & 1) value = __webpack_require__(value);
/******/ 		if(mode & 8) return value;
/******/ 		if((mode & 4) && typeof value === 'object' && value && value.__esModule) return value;
/******/ 		var ns = Object.create(null);
/******/ 		__webpack_require__.r(ns);
/******/ 		Object.defineProperty(ns, 'default', { enumerable: true, value: value });
/******/ 		if(mode & 2 && typeof value != 'string') for(var key in value) __webpack_require__.d(ns, key, function(key) { return value[key]; }.bind(null, key));
/******/ 		return ns;
/******/ 	};
/******/
/******/ 	// getDefaultExport function for compatibility with non-harmony modules
/******/ 	__webpack_require__.n = function(module) {
/******/ 		var getter = module && module.__esModule ?
/******/ 			function getDefault() { return module['default']; } :
/******/ 			function getModuleExports() { return module; };
/******/ 		__webpack_require__.d(getter, 'a', getter);
/******/ 		return getter;
/******/ 	};
/******/
/******/ 	// Object.prototype.hasOwnProperty.call
/******/ 	__webpack_require__.o = function(object, property) { return Object.prototype.hasOwnProperty.call(object, property); };
/******/
/******/ 	// __webpack_public_path__
/******/ 	__webpack_require__.p = "";
/******/
/******/
/******/ 	// Load entry module and return exports
/******/ 	return __webpack_require__(__webpack_require__.s = 0);
/******/ })
/************************************************************************/
/******/ ([
/* 0 */
/***/ (function(module, exports, __webpack_require__) {

module.exports = __webpack_require__(1);


/***/ }),
/* 1 */
/***/ (function(module, __webpack_exports__, __webpack_require__) {

"use strict";
__webpack_require__.r(__webpack_exports__);
function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _inheritsLoose(subClass, superClass) { subClass.prototype = Object.create(superClass.prototype); subClass.prototype.constructor = subClass; _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

/* harmony default export */ __webpack_exports__["default"] = (function (_ref) {
  var _class, _temp;

  var Component = _ref.Component,
      createElement = _ref.createElement,
      createRef = _ref.createRef;
  return _temp = _class = /*#__PURE__*/function (_Component) {
    _inheritsLoose(ReactHint, _Component);

    function ReactHint() {
      var _this;

      for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
        args[_key] = arguments[_key];
      }

      _this = _Component.call.apply(_Component, [this].concat(args)) || this;

      _defineProperty(_assertThisInitialized(_this), "_hint", createRef());

      _defineProperty(_assertThisInitialized(_this), "_container", createRef());

      _defineProperty(_assertThisInitialized(_this), "state", {
        target: null,
        hidden: true
      });

      _defineProperty(_assertThisInitialized(_this), "target", null);

      _defineProperty(_assertThisInitialized(_this), "placement", null);

      _defineProperty(_assertThisInitialized(_this), "_containerStyle", {
        position: 'relative'
      });

      _defineProperty(_assertThisInitialized(_this), "toggleEvents", function (_ref2, flag) {
        var events = _ref2.events,
            _ref2$events = _ref2.events,
            click = _ref2$events.click,
            focus = _ref2$events.focus,
            hover = _ref2$events.hover;
        var action = flag ? 'addEventListener' : 'removeEventListener';
        var hasEvents = events === true;
        (click || hasEvents) && document[action]('click', _this.toggleHint);
        (focus || hasEvents) && document[action]('focusin', _this.toggleHint);
        (hover || hasEvents) && document[action]('mouseover', _this.toggleHint);
        (click || hover || hasEvents) && document[action]('touchend', _this.toggleHint);
      });

      _defineProperty(_assertThisInitialized(_this), "toggleHint", function (_temp2) {
        var _ref3 = _temp2 === void 0 ? {} : _temp2,
            _ref3$target = _ref3.target,
            target = _ref3$target === void 0 ? null : _ref3$target;

        target = _this.getHint(target);
        clearTimeout(_this._timeout);
        _this._timeout = setTimeout(function () {
          _this.target = target;

          if (_this.target) {
            _this.placement = null;
          }

          _this.getHintData();
        }, target === null ? _this.props.delay.hide === undefined ? _this.props.delay : _this.props.delay.hide : _this.props.delay.show === undefined ? _this.props.delay : _this.props.delay.show);
      });

      _defineProperty(_assertThisInitialized(_this), "getHint", function (el) {
        var _this$props = _this.props,
            attribute = _this$props.attribute,
            persist = _this$props.persist;
        var target = _this.target;

        while (el) {
          if (el === document) break;
          if (persist && el === _this._hint.current) return target;
          if (el.hasAttribute(attribute)) return el;
          el = el.parentNode;
        }

        return null;
      });

      _defineProperty(_assertThisInitialized(_this), "shallowEqual", function (a, b) {
        var keys = Object.keys(a);
        return keys.length === Object.keys(b).length && keys.reduce(function (result, key) {
          return result && (typeof a[key] === 'function' && typeof b[key] === 'function' || a[key] === b[key]);
        }, true);
      });

      _defineProperty(_assertThisInitialized(_this), "getHintData", function () {
        if (!_this.target) {
          _this.setState({
            hidden: true
          });

          return;
        }

        var _this$props2 = _this.props,
            attribute = _this$props2.attribute,
            autoPosition = _this$props2.autoPosition,
            position = _this$props2.position;
        var content = _this.target.getAttribute(attribute) || '';
        var at = _this.placement || _this.target.getAttribute(attribute + "-at") || position;

        var _this$_container$curr = _this._container.current.getBoundingClientRect(),
            containerTop = _this$_container$curr.top,
            containerLeft = _this$_container$curr.left;

        var _this$_hint$current$g = _this._hint.current.getBoundingClientRect(),
            hintWidth = _this$_hint$current$g.width,
            hintHeight = _this$_hint$current$g.height;

        var _this$target$getBound = _this.target.getBoundingClientRect(),
            targetTop = _this$target$getBound.top,
            targetLeft = _this$target$getBound.left,
            targetWidth = _this$target$getBound.width,
            targetHeight = _this$target$getBound.height;

        if (autoPosition && !_this.placement) {
          var isHoriz = ['left', 'right'].includes(at);
          var _document$documentEle = document.documentElement,
              clientHeight = _document$documentEle.clientHeight,
              clientWidth = _document$documentEle.clientWidth;
          var directions = {
            left: (isHoriz ? targetLeft - hintWidth : targetLeft + (targetWidth - hintWidth >> 1)) > 0,
            right: (isHoriz ? targetLeft + targetWidth + hintWidth : targetLeft + (targetWidth + hintWidth >> 1)) < clientWidth,
            bottom: (isHoriz ? targetTop + (targetHeight + hintHeight >> 1) : targetTop + targetHeight + hintHeight) < clientHeight,
            top: (isHoriz ? targetTop - (hintHeight >> 1) : targetTop - hintHeight) > 0
          };

          if (!at || !directions[at]) {
            switch (at) {
              case 'left':
                if (!directions.left) at = 'right';
                if (!directions.top) at = 'bottom';
                if (!directions.bottom) at = 'top';
                break;

              case 'right':
                if (!directions.right) at = 'left';
                if (!directions.top) at = 'bottom';
                if (!directions.bottom) at = 'top';
                break;

              case 'bottom':
                if (!directions.bottom) at = 'top';
                if (!directions.left) at = 'right';
                if (!directions.right) at = 'left';
                break;

              case 'top':
              default:
                if (!directions.top) at = 'bottom';
                if (!directions.left) at = 'right';
                if (!directions.right) at = 'left';
                break;
            }
          }
        }

        var top, left;

        switch (at) {
          case 'left':
            top = targetHeight - hintHeight >> 1;
            left = -hintWidth;
            break;

          case 'right':
            top = targetHeight - hintHeight >> 1;
            left = targetWidth;
            break;

          case 'bottom':
            top = targetHeight;
            left = targetWidth - hintWidth >> 1;
            break;

          case 'top':
          default:
            top = -hintHeight;
            left = targetWidth - hintWidth >> 1;
        }

        if (hintHeight > 0 || hintWidth > 0) {
          // This prevents react-hint from rotating the placements 
          _this.placement = at;
        }

        var newState = {
          content: content,
          at: at,
          top: top + targetTop - containerTop | 0,
          left: left + targetLeft - containerLeft | 0,
          hidden: false
        };

        _this.setState(newState);
      });

      return _this;
    }

    var _proto = ReactHint.prototype;

    _proto.componentDidMount = function componentDidMount() {
      this.toggleEvents(this.props, true);
    };

    _proto.componentWillUnmount = function componentWillUnmount() {
      this.toggleEvents(this.props, false);
      clearTimeout(this._timeout);
    };

    _proto.componentDidUpdate = function componentDidUpdate() {
      if (this.target && !this.state.hidden) {
        this.getHintData();
      }
    };

    _proto.shouldComponentUpdate = function shouldComponentUpdate(props, state) {
      return !this.shallowEqual(state, this.state) || !this.shallowEqual(props, this.props);
    };

    _proto.render = function render() {
      var _this$props3 = this.props,
          className = _this$props3.className,
          onRenderContent = _this$props3.onRenderContent;
      var _this$state = this.state,
          content = _this$state.content,
          at = _this$state.at,
          top = _this$state.top,
          left = _this$state.left;
      return createElement("div", {
        ref: this._container,
        style: this._containerStyle
      }, createElement("div", {
        className: className + " " + className + "--" + at,
        ref: this._hint,
        role: "tooltip",
        style: {
          top: top,
          left: left,
          display: this.target ? undefined : 'none'
        }
      }, this.target && (onRenderContent ? onRenderContent(this.target, content) : createElement("div", {
        className: className + "__content"
      }, content))));
    };

    return ReactHint;
  }(Component), _defineProperty(_class, "defaultProps", {
    attribute: 'data-rh',
    autoPosition: false,
    className: 'react-hint',
    delay: 0,
    events: false,
    onRenderContent: null,
    persist: false,
    position: 'top'
  }), _temp;
});

/***/ })
/******/ ])["default"];
});

demo/src/index.js
import React from 'react'
import {render} from 'react-dom'
import ReactHintFactory from 'src'
import 'css/index.css'
import './index.css'

const ReactHint = ReactHintFactory(React)
class App extends React.Component {
  onRenderContent = (target, content) => {
    const {catId} = target.dataset
    const width = 240
    const url = `https://images.pexels.com/photos/${catId}/pexels-photo-${catId}.jpeg?w=${width}`

    return (
      <div className="custom-hint__content">
        <img src={url} width={width} />
        <button
          ref={(ref) => ref && ref.focus()}
          onClick={() => this.instance.toggleHint()}
        >
          Ok
        </button>
      </div>
    )
  }

  render() {
    return (
      <div>
        <ReactHint
          auto
          position="top"
          events={{hover: true}}
          onRenderContent={() => (
            <div
              style={{
                minWidth: '100px',
                padding: '5px',
                border: '1px solid black',
                zIndex: '99999',
              }}
            >
              Hello! Something long that could demonstrate the issue very
              easily. Infinite loop should not happen here. Infinite loop should
              not happen here
            </div>
          )}
          attribute="data-tip1"
        />
        <ReactHint autoPosition events delay={{show: 100, hide: 1000}} />
        <ReactHint
          persist
          attribute="data-custom"
          className="custom-hint"
          events={{click: true}}
          onRenderContent={this.onRenderContent}
          ref={(ref) => (this.instance = ref)}
        />

        <button data-tip1>Long text with content</button>
        <button data-rh="Default">Default</button>
        <button data-rh="This is a very very long text, This is a very very long text, This is a very very long text">
          Long text
        </button>
        <button data-rh="Top" data-rh-at="top">
          Top
        </button>
        <button data-rh="Right" data-rh-at="right">
          Right
        </button>
        <button data-rh="Bottom" data-rh-at="bottom">
          Bottom
        </button>
        <button data-rh="Left" data-rh-at="left">
          Left
        </button>

        <button data-custom data-custom-at="bottom" data-cat-id="10913">
          Click Me
        </button>

        <button data-custom data-custom-at="bottom" data-cat-id="416088">
          Click Me
        </button>
      </div>
    )
  }
}

render(<App />, demo)
